### ISSUE TITLE
Add the global redis cache the ability for caching query results

### ISSUE THREAD

**Issue by andreisucman** [parse-community/parse-server#8495]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
It would be nice to have global redis cache cache query results from Mongo DB by default.

### Steps to reproduce
1) Install a redis server and make sure it's running.
2) Connect the redis server to parse via redis adapter like this:

`const RedisCacheAdapter = require("parse-server").RedisCacheAdapter;`

```
const serverConfig = {
   ...
   cacheAdapter: new RedisCacheAdapter({ url: process.env.REDIS_URL }),
   ...
}
```

### Environment
node 18.2.0
redis-server on local docker v7.0.10
redis-server on digital ocean droplet with ubuntu 22 v7.0.4

Server
- Parse Server version: `^6.0.0`
- Operating system: `Win 64`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `Redis server tested both on redis-server version=7.0.10 on a local docker container as well as digital ocean droplet with ubuntu 22 redis-version 7.0.4`

Database
- System (MongoDB or Postgres): `Mongo`
- Database version: `5.0.15`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `Atlas`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `JavaScript`
- SDK version: `^4.0.1`

**Comment by mtrezza** [comment]
Could you please specify the exact versions in the environment section.

**Comment by mtrezza** [comment]
This sounds more like you're looking for support on how to set up Redis cache with Parse Server.

- For help with Parse Platform we recommend our [community forum](https://community.parseplatform.org) or our [community chat](https://chat.parseplatform.org)
- For general coding questions we recommend Stack Overflow using the [parse-platform](https://stackoverflow.com/questions/tagged/parse-platform) tag
- For network and server questions we recommend ServerFault using the [parse-server](https://serverfault.com/questions/tagged/parse-server) tag

**Comment by andreisucman** [comment]
Thank you for these resources I will check them out.

But there's not really any setup necessary, just copy and paste 2 lines of code which I did.

And my redis servers are up and running.

So this might be a bug rather than just me asking for support.

Thank you.

**Comment by mtrezza** [comment]
Thanks for providing more details, if you can confirm that this is a bug and not an issue with your custom setup then please post here and we'll reopen the issue.

**Comment by dblythy** [comment]
From my observation, Parse Server uses an inbuilt cache / redis cache to store **User Roles** and **User Data** to speed up the time it takes to resolve a cloud request (`beforeFind`, `afterFind`, etc). It does not provide a "global cache" that stores query responses.

If you use a Redis GUI such as RedisInsight you can observe this happening.

**Comment by andreisucman** [comment]
Thanks for confirming my surmise.

**Comment by dblythy** [comment]
Yes, so I think you would need to reword this as a feature request rather than a bug because the cache is working as expected. 

See: https://docs.parseplatform.org/parse-server/guide/#configuring-cache-adapters

> By default, parse-server provides an internal cache layer to speed up schema verifications, user, roles and sessions lookup.

> In some cases, in distributed environment, you may want to use a distributed cache like Redis.

**Comment by dblythy** [comment]
I think this is a duplicate of #4750.

I think it would be hard to decide how a global cache should work, if it's providing a short-circuited response from `beforeFind` and `afterFind` triggers. A developer might have custom logic that they want to run in a trigger, regardless of whether the response is cached or not.

It seems like Flo's suggestion was:

1. Allow beforeFind triggers to return objects
- This could allow for custom query returns, for example if a developer is using a custom search such as ElasticSearch
- This could allow cached objects to be returned

2. Pass objects in `afterFind` to a cache adapter

It's worth noting that **whether or not** a `cacheAdapter` is specified, Parse Server will cache schema verifications, user, roles and sessions lookup in an in-memory cache. If we introduce the ability to cache query responses, we will need to make sure that the default behaviour is considered.

We could possibly add a parameter to `SCHEMA` or `Cloud Validators` of `cache: true` which interprets whether to cache query results with the `CacheAdapters` `ttl`

**Comment by andreisucman** [comment]
Just as we were speaking there was a package out that does query result caching: 

https://community.parseplatform.org/t/parse-cache-memory-a-cache-library-customized-for-parse-server-queries/3367/5

I've tested it and my query responses improved from 200ms to 1ms.

**Comment by mtrezza** [comment]
Given the significant cost saving such a feature could bring, let's revive this feature request. Let's start with a minimal feature scope to implement in Parse Server, since it already has the cacheAdapter interface.

Scope:

- Maintain all Cloud Code triggers to be called agnostically, regardless of whether the data comes from the DB or cache.
- Allow to configure cache params (TTL) per query request for simplicity and max. versatility.
- A data-use driven strategy where a query gets an ID assigned; for example to cache 2 queries for the same data / class with different TTLs
- Add methods to flush the cache (a) for the entire query ID, (b) for a entry bases on a filter
